apps/src/applab/assetManagement/AssetUploader.jsx

```diff
@@ -0,0 +1,70 @@
+
+/**
+ * A component for managing hosted assets.
+ */
+module.exports = React.createClass({
+  propTypes: {
+    onUploadStart: React.PropTypes.func.isRequired,
+    onUploadDone: React.PropTypes.func.isRequired,
+    channelId: React.PropTypes.string.isRequired,
+    typeFilter: React.PropTypes.string,
+    uploadsEnabled: React.PropTypes.bool.isRequired
+  },
+
+  componentDidMount: function () {
+    var props = this.props;
+
+    $(React.findDOMNode(this.refs.uploader)).fileupload({
+      dataType: 'json',
+      url: '/v3/assets/' + props.channelId + '/new',
+      // prevent fileupload from replacing the input DOM element, which
+      // React does not like
+      replaceFileInput: false,
+      add: function (e, data) {
+        props.onUploadStart();
+        data.submit();
+      },
+      done: function (e, data) {
+        props.onUploadDone(data.result);
+      },
+      error: function (e, data) {
+        props.onUploadError(e.status);
+      }
+    });
+  },
+
+  componentWillUnmount: function () {
+    $(React.findDOMNode(this.refs.uploader)).fileupload('destroy');
+  },
+
+  /**
+   * We've hidden the <input type="file"/> and replaced it with a big button.
+   * Forward clicks on the button to the hidden file input.
+   */
+  fileUploadClicked: function () {
+    var uploader = React.findDOMNode(this.refs.uploader);
+    uploader.click();
+  },
+
+  render: function () {
+    // NOTE: IE9 will ignore accept, which means on this browser we can end
+    // up uploading files that dont match typeFilter
+    return (
+      <span>
+        <input
+            ref="uploader"
+            type="file"
+            style={{display: 'none'}}
+            accept={(this.props.typeFilter || '*') + '/*'}/>
```

Comment:
There are some cases where we want to restrict this to a particular type (i.e. I'm adding an image to a background) and others where it can be anything (I just clicked the gear icon). In IE9 we'll never restrict, even in case 1.